Tutorials/Style/Functions-Docstrings.MD

```diff
@@ -0,0 +1,169 @@
+
+returns
+
+```
+Help on built-in function len in module builtins:
+
+len(obj, /)
+    Return the number of items in a container.
+```
+
+This tells us that `len` is a built in function that comes with python, and that it takes one input (an object) and returns the number of items in that container object. 
+
+`help()` also works on other objects, such as variables, which can be useful if you aren't sure what a particular variable is. For instance:
+
+```python
+x = [1, 2]
+help(x)
+```
+
+tells us that `x` is a list, and gives us all sorts of information about the sorts of methods and functions that can be used with lists. 
+
+___
+
+## Docstrings
+
+What happens if you write a function for your own code and then make it publicly available for other people to use? How will users know what your function does if they need to understand it or adapt it? Well, you could comment your code (always good practice), but for functions (and classes) in particular, you can use the magic of **_docstrings_**. 
+
+### How do I write a docstring?
+
+It's easy to write a docstring. Just under your function definition, add triple quotes ` ''' `, write your docstring, and then end the docstring with another set of triple quotes ` ''' `. For example: 
+
+```python
+def square(x):
+    '''This is my docstring'''
+    return x**2
+```
+
+Now that we have our docstring, it will show up when we run `help()` on our function! This is a major advantage of using docstrings instead of comments when documenting functions. For instance: 
+
+```python
+def square(x):
+    '''This is my docstring'''
+    return x**2
+
+help(square)
+```
+
+returns
+
+```
+Help on function square in module __main__:
+
+square(x)
+    This is my docstring
+
+```
+
+which tells us that the function `square` is defined in the main code, it takes one input `x`, and that... it has a docstring that says `this is my docstring`. That's a docstring alright, but it isn't a very helpful one...
+
+### How do I write a _good_ docstring?
+
+It is only slightly harder to write a good docstring than it is to write a docstring. Best practices for writing a docstrings are: 
+- Describe what the function _does_ (this is usually the first line)
+- Describe what the function takes as an input, including data types, formats, variable names, and units if applicable
+- Describe what the function gives as an output, includeing data types, formats, and units if applicable
+- Note any domains over which the function will/will not work (for instance: does this work for integers only? Positive numbers?)
+- Note any dependencies the function has on other packages, functions, object types, etc. 
+Note that with the triple quotes, you can write a docstring that spans multiple lines, so if this seems like a lot to include, you have as much room as you need. 
+
+Here is how I would document the `square` function:
+
+```python
+def square(x):
+    '''
+    Squares the input. 
+    
+    The square function takes a single input:
```

Comment:
I think it would be worthwhile to do a brief description of standard docstring styles (http://queirozf.com/entries/docstrings-by-example-documenting-python-code-the-right-way) and advocate for a consistent way of listing the parameters, returns, and examples (I'm partial to the numpy style, more examples here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy). Using an accepted style will make docstrings easier to read and also easier for automatic documentation software to parse in a smart way, so it's a good habit to teach to new documentors!